pkg/policy/l4.go

```diff
@@ -256,7 +256,28 @@ func (l4 *L4Filter) ToMapState(direction trafficdirection.TrafficDirection) MapS
 		TrafficDirection: direction.Uint8(),
 	}
 
+	// find the L7 rules for the wildcard entry, if any
+	var wildcardL7Policy *PerSelectorPolicy
+	if l4.wildcard != nil {
+		wildcardL7Policy = l4.L7RulesPerSelector[l4.wildcard]
+	}
+
 	for cs, l7 := range l4.L7RulesPerSelector {
+		// Skip generating L3/L4 keys if L4-only key (for the same L4 port and
+		// protocol) has the same effect w.r.t. redirecting to the proxy or not,
+		// considering that L3/L4 key should redirect if L4-only key does. In
+		// summary, if have both L3/L4 and L4-only keys:
+		//
+		// L3/L4        L4-only      Skip generating L3/L4 key
+		// redirect     no redirect  no   (this case tested below)
+		// no redirect  no redirect  yes  (same effect)
```

Comment:
Does `l4.L7RulesPerSelector` have an entry for these cases where there is no redirect? ie are we actually handling that case in this loop?

Reply:
Technically I think we could still install the l3/l4 key in this case and it would save one map lookup for packets that match the l3/l4 policy on the dataplane, but we can probably optimize this later.

Reply:
As one of the changes in this PR overall, `l4.L7RulesPerSelector` now has an entry also for non-redirects, with `nil` value.

On the second point, yes we can do that, but the effect is now smaller due to the change in the lookup order in the datapath.